File: v1/Cpp/boj/Beginner/01_Sorting/QuickSort.cpp
//
// Created by seungsu on 22. 11. 17.
//

#include <iostream>
#include <vector>

using namespace std;

static void quick_sort(vector<int>& v, int start, int end) {
    // 종료 조건 : 원소의 개수가 1개인 경우
    if (start >= end)
        return;

    int pivot = start;
    int left = start + 1;
    int right = end;

    while (left <= right) {
        while (v[pivot] >= v[left] && left <= end)
            left++;

        while (v[pivot] <= v[right] && right > start)
            right--;

        if (left > right)
            std::swap(v[pivot], v[right]);
        else
            std::swap(v[left], v[right]);
    }

    quick_sort(v, 0, right - 1);
    quick_sort(v, right + 1, end);
}

void quick_sort_test() {
    int n;
    cin >> n;

    vector<int> v(n);
    for (int i = 0; i < n; i++)
        cin >> v[i];

    quick_sort(v, 0, n - 1);

    cout << "[ 정렬 완료 ]" << "\n";
    for (int val : v)
        cout << val << " ";
}

File: v1/Java/boj/Beginner/A_Sorting/QuickSort.java
package boj.Beginner.A_Sorting;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class QuickSort {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int[] arr;
    static int idx = 0;

    private static void quick_sort(int[] arr, int start, int end) {
        if (start >= end) return;       // 종료 조건: 원소의 개수가 1

        int pivot = start;
        int left = start + 1;
        int right = end;

        while (left <= right) {
            // left: pivot보다 큰 값 찾기
            while (left <= end && arr[left] <= arr[pivot])
                left++;

            // right: pivot보다 작은 값 찾기
            while (right > start && arr[right] >= arr[pivot])
                right--;

            int tmp = arr[right];
            if (left > right) {                         // 만약 엇갈렸다면 작은 데이터(right)와 피벗 교환
                arr[right] = arr[pivot];
                arr[pivot] = tmp;
            }
            else {                                      // 엇갈리지 않았다면 작은 데이터와 큰 데이터 교체
                arr[right] = arr[left];
                arr[left] = tmp;
            }
        }

        // 분할 이후 왼쪽 부분, 오른쪽 부분에서 각각 다시 정렬을 수행 => arr[right]에는 피봇값
        quick_sort(arr, start, right - 1);
        quick_sort(arr, right + 1, end);
    }

    public static void main(String[] args) throws IOException {
        arr = new int[10];
        st = new StringTokenizer(br.readLine());

        while (st.hasMoreTokens())
            arr[idx++] = Integer.parseInt(st.nextToken());

        quick_sort(arr, 0, arr.length - 1);

        System.out.println("[ 퀵정렬 수행 결과 ]");
        for (int a : arr)
            System.out.print(a + " ");
    }
}
